libp2p/crypto/chacha20poly1305.nim

```diff
@@ -0,0 +1,91 @@
+## Nim-Libp2p
+## Copyright (c) 2020 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+## This module integrates BearSSL ChaCha20+Poly1305
+##
+## This module uses unmodified parts of code from
+## BearSSL library <https://bearssl.org/>
+## Copyright(C) 2018 Thomas Pornin <pornin@bolet.org>.
+
+# RFC @ https://tools.ietf.org/html/rfc7539
+
+import bearssl
+
+const
+  ChaChaPolyKeySize = 32
+  ChaChaPolyNonceSize = 12
+  ChaChaPolyTagSize = 16
+  
+type
+  ChaChaPoly* = object
+  ChaChaPolyKey* = array[ChaChaPolyKeySize, byte]
+  ChaChaPolyNonce* = array[ChaChaPolyNonceSize, byte]
+  ChaChaPolyTag* = array[ChaChaPolyTagSize, byte]
+
+converter toChaChaPolyKey*(s: seq[byte]): ChaChaPolyKey =
+  assert s.len == ChaChaPolyKeySize
+  copyMem(addr result[0], unsafeaddr s[0], ChaChaPolyKeySize)
+
+converter toChaChaPolyNonce*(s: seq[byte]): ChaChaPolyNonce =
+  assert s.len == ChaChaPolyNonceSize
+  copyMem(addr result[0], unsafeaddr s[0], ChaChaPolyNonceSize)
+
+converter toChaChaPolyTag*(s: seq[byte]): ChaChaPolyTag =
+  assert s.len == ChaChaPolyTagSize
+  copyMem(addr result[0], unsafeaddr s[0], ChaChaPolyTagSize)
+  
+# bearssl allows us to use optimized versions
+# this is reconciled at runtime
+# we do this in the global scope / module init
+
+# try for the best first
+var
+  chachapoly_native_impl: Poly1305Run = poly1305CtmulqGet()
```

Comment:
Please avoid using global variables, bearssl algorithm implementation version can be controlled via compile-time options.

Reply:
I will have a look if I can find those options.
Any reason why using a internal global is bad tho? 
This is a pretty decent language feature, specially when used without any particular memory pressure or so (in this case kinda a lazy linking/jit)

Reply:
This will possibly break with threads, you can make it thread local to avoid issues. In general unless there is a really good reason I would stay away from globals.

Reply:
Any thread will be used after global initialization, this is just loading one function pointer and cache it for the whole program execution to avoid to fetch it every-time.
Access to global statics should be guarded only if mutable (it is in this case but internally never mutated).
Anyway I'll do as asked.

Reply:
I agree, it doesn't seem like it would be an issue here and I'm sure you know the perils of globals 😸.

Reply:
in addition to all the usual suspects (reasonability, readability, etc etc), a downside of globals is the side-effect-on-import that they cause, effectively disabling dead code elimination, making executables larger (because people are lazy and overimport).

would a `let` or `const` be enough?

Reply:
ah, good point about dead code elimination...

Reply:
In this case had to be a `var` cos was lazy linking, curve had a `let`.
I removed all usage anyway.
My intention was caching due to the fact that we don't know what bearssl does within their calls and caching helps in case they do heavy lifting (unlikely) or even unsafe code (unlikely cos OSS library and I checked). 
But since we could not agree I just follow the way bearssl was used until now. Can be fixed later anyway if we figure that in fact it was a bottleneck so no biggie either way.